sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -4034,3 +4009,35 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
     if (targetLength > charLength) StringRPad(expr, Literal(targetLength)) else expr
   }
 }
+
+/**
+ * This rule removes metadata columns from `DataSourceV2Relation` under 2 cases:
+ *   - A single v2 scan (can be produced by `spark.table`), which is similar to star expansion, and
+ *     metadata columns should only be picked by explicit references.
+ *   - V2 scans under writing commands, as we can't insert into metadata columns.
```

Comment:
This is for the `table` in `InsertIntoStatement`. How about the `query`? E.g. `spark.table(...).write.insertInto(...)`. Do we need to remove metadata columns for the query here if it also is a v2 scan?